### DISCUSSION TITLE
Docker & translation: Server won't append new translation

### DISCUSSION THREAD

**Discussion by n1smithy** [Ylianst/MeshCentral#6461] [Q&A] [answered]
Hi there,

I tried to improve the German translation. So I used the html-based translation-editor and tried to append the new translations. But somehow the (official) docker-image won't use the new translations.

Is there anything special I must pay attention?

**Comment by si458** [comment]
First check inside the `meshcentral-data` folder, do u see a translation.json file?
If not, u need to click the Save button first in the Web translate page, then click translate button

**Reply by n1smithy** [reply]
Hi Simon!

The file is there.

**Reply by si458** [reply]
just tried it here and it seems to work?
1. login to meshcentral
2. visit `https://mydomain.com/translator.htm`
3. select you language, select your value, enter translated text in bottom field, click set
4. once finished, click `save to server`, then click `translate to server`
5. wait 5(ish)mins (have a tea 🫖 or coffee ☕)
6. restart meshcentral container
7. hard refresh your webpage

**Comment by n1smithy** [comment]
There must be something wrong. My server runs into overload when translating the translation. I logged into the meshcentral-docker-container and run `/opt/meshcentral/meshcentral/translate/node translate.js translateall` manually, and the process was being killed when doing "Translating HTML (fr): invite.handelbars".

**Reply by n1smithy** [reply]
Deleted the server-created translate.json-file and made some changes by the steps you posted. But after re-up the docker-container there are no changes. Weird!

**Reply by si458** [reply, marked as answer]
> the process was being killed when doing "Translating HTML (fr): invite.handelbars".

self explanatory: you are running out of memory, the translate tool can use upto 4GB of memory,
so increase the memory for your meshcentral container OR the VM itself

**Reply by n1smithy** [reply]
Oh boy, any I already made an upgrade for my server. But there are several machines running on it. Thanks for your help!!!

**Reply by n1smithy** [reply]
Oh, and what I forgot: Yes, this was the solution. Translation now works!

**Comment by n1smithy** [comment]
There's something I didn't mention: My docker-containers run with remapped UIDs and GIDs. May be this important?

EDIT: not important

**Comment by n1smithy** [comment]
It happens again! After I made a lot of improvements within the translation the server-translation-process does not finish (or appends the translations). And I have an idea what could be the problem:
![grafik](https://github.com/user-attachments/assets/bad80257-ce8f-42f5-a1e3-cc31cb12ed05)

As you can see, sometimes the special characters will be translated correctly and sometimes not. After I edited and set a line with a not-translated character-code it will be shown as the right icon within my translation.
![grafik](https://github.com/user-attachments/assets/a8b50d82-e7a3-4833-876b-b23b072524b3)

I have the suspicion that this could crash the translation-process. :-/

**Reply by si458** [reply]
Again, make sure u have memory to do translations,
my server has 4 cores and 8GB of ram to complete correctly

Also, special characters need to be the html codes and not the actual symbol themselves from what I'm aware! I would need to double-check that

**Comment by n1smithy** [comment]
I know the reason know why my translations dissappeared after restarting my containers.

I always used "docker-compose **down**" which seems to delete the translated websites.

So if one want to test his translations, do the following:

`docker-compose stop`
and then again
`docker-compose up -d`

After that it worked. Damn …

**Reply by si458** [reply]
Ah right yes!
That's because it stores the translated pages inside `/opt/meshcentral/MeshCentral/views/translations`

To make them persistent,
1. create a volume and map it to `/opt/meshcentral/meshcentral-web`
2. Run ur translate, etc
3. Copy the folder `translations` into `/opt/meshcentral/meshcentral-web/views/`
4. Docker compose down/up
5. It will then always use the files from the `meshcentral-web` folder!

You just have to remember every time u wanna do translations to do step 2 and 3

OR if u are happy with ur translations, upload ur `translations.json` file here or email them to myself 
And we can add them into the official docker image for new releases or the docker `master` image :)

**Reply by n1smithy** [reply]
I've made a script using `docker exec -t (…)` for updating and restarting the containers. Works well! :-D

When I'm feeling well with my translations I'm going to send them to you, Simon. Thanks for your valuable help!

**Reply by si458** [reply]
No worries glad i could help!
Don't forget to donate if u can ❤️ https://www.si458.co.uk/2024/01/05/donation/
